design-docs/java-annotation-processing.md

```diff
@@ -44,74 +48,125 @@ public class CompileOptions extends AbstractOptions {
  * a non-empty `processorpath` adds a `-processorpath` argument with `processorpath.asPath` as value
  * a non-empty `processors` list adds a `-processor` argument with the `processors` joined with a `,` as value
  * each entry in `processorArgs` is added as `-Akey=value` arguments
- * `proc` with a value of `none` is added as a `-proc:none` argument
- * `proc` with a value of `only` is added as a `-proc:only` argument
+ * `annotationProcessing` set to `false` is added as a `-proc:none` argument
  * Add integration test with an annotation processor in the `compileOnly` configuration and an empty `compileJava.processorpath`, and verify that it ran (checks non-regression)
  * Add integration test with an annotation processor in the `compileJava.options.processorpath` and another in the `compileOnly` configuration, and verify that the former ran but the latter didn't
 
-### Story - Add processor path to `SourceSet` (TBD if this should live in gradle core)
+### Story - Allow specification of generated sources destination directory for compile tasks
+
+This is important for IDEs and for other tools that process source files (e.g. transpiling Java to JavaScript or ObjectiveC)
+
+Add `generatedSourcesDestinationDir` (similar to `destinationDir`) to `AbstractCompile`, as a `File`: when non-null, it is passed as a `-s` argument to the Java compiler (including for Groovy or Scala compilations).
+
+The `clean` task (dynamic `cleanCompileJava` task et al.) deletes the `generatedSourcesDestinationDir`.
+
+Non-incremental compilation should clean the `generatedSourcesDestinationDir` just like it cleans `destinationDir`.
+
+Note: incremental compilation should track generated sources and try to delete stale ones, but that's an incremental compilation issue, as it also has to track generated classes, independently of how annotation processors are configured.
+
+#### Backwards compatibility
+
+ * The new arguments must be added before the `compilerArgs` so that user-specified arguments in `compilerArgs` override the ones added by those new properties (similar to how `compilerArgs` can override `sourcepath` added in Gradle 2.4, or any other compiler argument
+
+#### Test cases
+
+ * The default behavior doesn't change (no change in existing tests, e.g. in `JavaCompilerArgumentsBuilderTest` and `CommandLineJavaCompilerArgumentsGeneratorTest`)
+ * a non-null `generatedSourcesDestinationDir` adds a `-s` argument
+ * The `clean` task (dynamic `cleanCompileJava` task et al.) deletes the `generatedSourcesDestinationDir`
+ * Add integration test with an annotation processor and verify that it generates source files in the `generatedSourcesDestinationDir`
+ * Add integration test with an annotation processor, compile, delete a source file, and verify that the associated generated source file has been removed in `generatedSourcesDestinationDir`
+
+### Story - Add properties to `SourceSet` and `SourceSetOutput`
```

Comment:
See earlier comments wrt configuration of IDEs. I don't really care if it's in the java or java-base plugin, or antoher specific plugin, but it should IMO come with Gradle and not as a third-party, or the details likely won't be properly exposed on the Tooling API for IDEs.

Fortunately, SourceSet and SourceSetOutput can be enriched using _conventions_, so those additional properties could come from/with a specific plugin rather than the java-base plugin.